File: AnkiDroid/src/test/java/com/ichi2/utils/FileUtilTest.java
/*
 *  Copyright (c) 2020 David Allison <davidallisongithub@gmail.com>
 *
 *  This program is free software; you can redistribute it and/or modify it under
 *  the terms of the GNU General Public License as published by the Free Software
 *  Foundation; either version 3 of the License, or (at your option) any later
 *  version.
 *
 *  This program is distributed in the hope that it will be useful, but WITHOUT ANY
 *  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
 *  PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License along with
 *  this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.ichi2.utils;

import com.ichi2.async.ProgressSenderAndCancelListener;
import com.ichi2.compat.CompatHelper;

import org.junit.Assert;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TemporaryFolder;
import org.robolectric.annotation.Config;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Map;

import androidx.annotation.Nullable;

import static org.acra.util.IOUtils.writeStringToFile;
import static org.hamcrest.CoreMatchers.nullValue;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.core.Is.is;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertThrows;
import static org.junit.Assert.assertTrue;

public class FileUtilTest {

    @Rule
    public TemporaryFolder temporaryFolder = new TemporaryFolder();
    long mTestFolderSize;

    private static class DummyListener implements ProgressSenderAndCancelListener<Integer> {
        @Override
        public boolean isCancelled() {
            return false;
        }

        @Override
        public void doProgress(@Nullable Integer value) { }
    }

    private File addUnexpectedFileToDestination(File unsuccessfulDestDir) throws Exception {
        // Add an 'unexpected' file to destination directory with the same name and relative path as a directory in
        // the source directory
        // This triggers an exception
        File parentOfUnexpectedFile = new File(unsuccessfulDestDir.getAbsoluteFile() + File.separator + "parent"
                + File.separator + "child");
        if (!parentOfUnexpectedFile.mkdirs()) {
            throw new Exception("Directories could not be created inside empty directory '" + unsuccessfulDestDir + "'");
        }
        File unexpectedFile = new File(parentOfUnexpectedFile, "grandChild");
        if (!unexpectedFile.createNewFile()) {
            throw new Exception("File could not be created inside empty directory '" + parentOfUnexpectedFile + "'");
        }

        return unexpectedFile;
    }

    private File createSrcFilesForTest(File temporaryRoot, String testDirName) throws Exception {
        File grandParentDir = new File(temporaryRoot, testDirName);
        File parentDir = new File(grandParentDir, "parent");
        File childDir = new File(parentDir, "child");
        final File childDir2 = new File(parentDir, "child2");
        final File grandChildDir = new File(childDir, "grandChild");
        final File grandChild2Dir = new File(childDir2, "grandChild2");

        ArrayList<File> files = new ArrayList<>();
        files.add(new File(grandParentDir, "file1.txt"));
        files.add(new File(parentDir, "file2.txt"));
        files.add(new File(childDir, "file3.txt"));
        files.add(new File(childDir2, "file4.txt"));
        files.add(new File(grandChildDir, "file5.txt"));
        files.add(new File(grandChildDir, "file6.txt"));

        grandChildDir.mkdirs();
        grandChild2Dir.mkdirs();

        for (int i = 0; i < files.size(); ++i) {
            final File file = files.get(i);
            writeStringToFile(file, "File " + (i + 1) + " called " + file.getName());
            mTestFolderSize += file.length();
        }

        return grandParentDir;
    }

    @Config(sdk = { 21, 26 })
    @Test
    public void testCopyDirectory() throws Exception {
        // Create temporary root directory for holding test directories
        File temporaryRootDir = temporaryFolder.newFolder("tempRootDir");

        // Test for successful copy directory operation
        File srcDir = createSrcFilesForTest(temporaryRootDir, "srcDir");
        File successfulDestDir = new File(temporaryRootDir, "successfulDest");
        CompatHelper.getCompat().copyDirectory(srcDir, successfulDestDir, new DummyListener(), false);
        compareDirs(srcDir, successfulDestDir);


        // Test for unsuccessful copy directory operation
        File unsuccessfulDestDir = new File(temporaryRootDir, "failedDest");
        File unexpectedFile = addUnexpectedFileToDestination(unsuccessfulDestDir);
        org.junit.Assert.assertThrows(IOException.class, () -> CompatHelper.getCompat().copyDirectory(srcDir,
                unsuccessfulDestDir, new DummyListener(), false));

        // Test for successful copy directory operation after partial operation
        // by removing the unexpected file from the destination directory and trying again
        Assert.assertTrue(unexpectedFile.delete());
        CompatHelper.getCompat().copyDirectory(srcDir, unsuccessfulDestDir, new DummyListener() ,false);
        compareDirs(srcDir, unsuccessfulDestDir);
    }

    @Config(sdk = { 21, 26 })
    @Test
    public void testMoveDirectory() throws Exception {
        // Create temporary root directory for holding test directories
        File temporaryRootDir = temporaryFolder.newFolder("tempRootDir");

        // Test for successful move directory operation
        File srcDirToBeMovedSuccessfully = createSrcFilesForTest(temporaryRootDir, "srcDirToBeMovedSuccessfully");
        File srcDirForComparison = createSrcFilesForTest(temporaryRootDir, "srcDirForComparison");
        File successfulDestDir = new File(temporaryRootDir, "successfulDestDir");
        CompatHelper.getCompat().moveDirectory(srcDirToBeMovedSuccessfully, successfulDestDir, new DummyListener());
        compareDirs(srcDirForComparison, successfulDestDir);

        // Test for unsuccessful move directory operation
        File srcDirToBeMovedUnsuccessfully = createSrcFilesForTest(temporaryRootDir, "srcDirToBeMovedUnsuccessfully");
        File unsuccessfulDestDir = new File(temporaryRootDir, "unsuccessfulDestDir");
        File unexpectedFile = addUnexpectedFileToDestination(unsuccessfulDestDir);
        org.junit.Assert.assertThrows(IOException.class, () -> CompatHelper.getCompat().moveDirectory(srcDirToBeMovedUnsuccessfully,
                unsuccessfulDestDir, new DummyListener()));

        // Test for successful move directory operation after partial operation
        // by removing the unexpected file from the destination directory and trying again
        Assert.assertTrue(unexpectedFile.delete());
        CompatHelper.getCompat().moveDirectory(srcDirToBeMovedUnsuccessfully, unsuccessfulDestDir, new DummyListener());
        compareDirs(srcDirForComparison, unsuccessfulDestDir);
    }

    private void compareDirs(File srcDir, File destDir) {
        org.junit.Assert.assertTrue(srcDir.isDirectory());
        org.junit.Assert.assertTrue(destDir.isDirectory());

        final File[] srcFiles = srcDir.listFiles();
        org.junit.Assert.assertNotNull(srcFiles);

        for (final File srcFile : srcFiles) {
            final File destFile = new File(destDir, srcFile.getName());
            org.junit.Assert.assertTrue("src: " + srcFile + ", dest: " + destFile, destFile.exists());

            if (srcFile.isDirectory()) {
                org.junit.Assert.assertTrue(destFile.isDirectory());
                compareDirs(srcFile, destFile);
            } else {
                org.junit.Assert.assertFalse(destFile.isDirectory());
                Assert.assertEquals(srcFile.length(), destFile.length());
            }
        }
    }

    @Test
    public void testDirectorySize() throws Exception {
        // Create temporary root directory for holding test directories
        File temporaryRootDir = temporaryFolder.newFolder("tempRootDir");

        // Test for success scenario
        File dir = createSrcFilesForTest(temporaryRootDir, "dir");
        assertEquals(FileUtil.getDirectorySize(dir), mTestFolderSize);

        // Test for failure scenario by passing a file as an argument instead of a directory
        assertThrows(IOException.class, () -> FileUtil.getDirectorySize(new File(dir, "file1.txt")));
    }

    @Test
    public void ensureFileIsDirectoryTest() throws Exception {
        // Create temporary root directory for holding test directories
        File temporaryRootDir = temporaryFolder.newFolder("tempRootDir");

        // Create test data
        File testDir = createSrcFilesForTest(temporaryRootDir, "testDir");

        // Test for file which exists but isn't a directory
        org.junit.Assert.assertThrows(IOException.class, () -> FileUtil.ensureFileIsDirectory(new File(testDir, "file1.txt")));

        // Test for file which exists and is a directory
        FileUtil.ensureFileIsDirectory(new File(testDir, "parent"));

        // Test for directory which doesn't exist, but can be created
        FileUtil.ensureFileIsDirectory(new File(testDir, "parent2"));

        // Test for directory which doesn't exist, and cannot be created
        Assert.assertThrows(IOException.class, () -> FileUtil.ensureFileIsDirectory(
                new File(testDir.getAbsolutePath() + File.separator + "file1.txt"
                        + File.separator + "impossibleDir")));
    }

    @Test
    public void listFilesTest() throws Exception {
        // Create temporary root directory for holding test directories
        File temporaryRootDir = temporaryFolder.newFolder("tempRootDir");

        // Create valid input
        File testDir = createSrcFilesForTest(temporaryRootDir ,"testDir");
        ArrayList<File> expectedChildren = new ArrayList<>();
        expectedChildren.add(new File(testDir, "parent"));
        expectedChildren.add(new File(testDir, "file1.txt"));

        File[] testDirChildren = FileUtil.listFiles(testDir);

        // Check that listFiles lists all files in the directory
        for (final File testDirChild : testDirChildren) {
             assertTrue(expectedChildren.contains(testDirChild));
        }
        assertEquals(expectedChildren.size(), testDirChildren.length);

        // Create invalid input
        assertThrows(IOException.class, () -> FileUtil.listFiles(new File(testDir, "file1.txt")));
    }

    @Test
    public void testFileNameNull() {
        assertThat(FileUtil.getFileNameAndExtension(null), nullValue());
    }

    @Test
    public void testFileNameEmpty() {
        assertThat(FileUtil.getFileNameAndExtension(""), nullValue());
    }

    @Test
    public void testFileNameNoDot() {
        assertThat(FileUtil.getFileNameAndExtension("abc"), nullValue());
    }

    @Test
    public void testFileNameNormal() {
        Map.Entry<String, String> fileNameAndExtension = FileUtil.getFileNameAndExtension("abc.jpg");
        assertThat(fileNameAndExtension.getKey(), is("abc"));
        assertThat(fileNameAndExtension.getValue(), is(".jpg"));
    }

    @Test
    public void testFileNameTwoDot() {
        Map.Entry<String, String> fileNameAndExtension = FileUtil.getFileNameAndExtension("a.b.c");
        assertThat(fileNameAndExtension.getKey(), is("a.b"));
        assertThat(fileNameAndExtension.getValue(), is(".c"));
    }
}

File: AnkiDroid/src/test/java/com/ichi2/utils/FileUtilTest.kt
/*
 *  Copyright (c) 2020 David Allison <davidallisongithub@gmail.com>
 *
 *  This program is free software; you can redistribute it and/or modify it under
 *  the terms of the GNU General Public License as published by the Free Software
 *  Foundation; either version 3 of the License, or (at your option) any later
 *  version.
 *
 *  This program is distributed in the hope that it will be useful, but WITHOUT ANY
 *  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
 *  PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License along with
 *  this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.ichi2.utils

import com.ichi2.testutils.common.assertThrows
import org.acra.util.IOUtils.writeStringToFile
import org.hamcrest.CoreMatchers.nullValue
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.core.IsEqual.equalTo
import org.junit.Assume.assumeTrue
import org.junit.Rule
import org.junit.Test
import org.junit.rules.TemporaryFolder
import java.io.File
import java.io.IOException
import java.nio.file.FileSystems
import java.nio.file.Files
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertTrue
import kotlin.test.fail

class FileUtilTest {
    @get:Rule
    var temporaryDirectory = TemporaryFolder()
    private var testDirectorySize: Long = 0

    @Throws(Exception::class)
    private fun createSrcFilesForTest(
        temporaryRoot: File,
        testDirName: String,
    ): File {
        val grandParentDir = File(temporaryRoot, testDirName)
        val parentDir = File(grandParentDir, "parent")
        val childDir = File(parentDir, "child")
        val childDir2 = File(parentDir, "child2")
        val grandChildDir = File(childDir, "grandChild")
        val grandChild2Dir = File(childDir2, "grandChild2")
        val files =
            listOf(
                File(grandParentDir, "file1.txt"),
                File(parentDir, "file2.txt"),
                File(childDir, "file3.txt"),
                File(childDir2, "file4.txt"),
                File(grandChildDir, "file5.txt"),
                File(grandChildDir, "file6.txt"),
            )
        grandChildDir.mkdirs()
        grandChild2Dir.mkdirs()
        files.forEachIndexed { i, file ->
            writeStringToFile(file, "File " + (i + 1) + " called " + file.name)
            this.testDirectorySize += file.length()
        }
        return grandParentDir
    }

    @Test
    @Throws(Exception::class)
    fun listFilesTest() {
        // Create temporary root directory for holding test directories
        val temporaryRootDir = temporaryDirectory.newFolder("tempRootDir")

        // Create valid input
        val testDir = createSrcFilesForTest(temporaryRootDir, "testDir")
        val expectedChildren = ArrayList<File>()
        expectedChildren.add(File(testDir, "parent"))
        expectedChildren.add(File(testDir, "file1.txt"))

        val testDirChildren = FileUtil.listFiles(testDir)

        // Check that listFiles lists all files in the directory
        for (testDirChild in testDirChildren) {
            assertTrue(expectedChildren.contains(testDirChild))
        }
        assertEquals(expectedChildren.size.toLong(), testDirChildren.size.toLong())

        // Create invalid input
        assertThrows<IOException> { FileUtil.listFiles(File(testDir, "file1.txt")) }
    }

    @Test
    fun testFileNameEmpty() {
        assertThat(FileNameAndExtension.fromString(""), nullValue())
    }

    @Test
    fun testFileNameNoDot() {
        assertThat(FileNameAndExtension.fromString("abc"), nullValue())
    }

    @Test
    fun testFileNameNormal() {
        val (fileName, extension) = getValidFileNameAndExtension("abc.jpg")
        assertThat(fileName, equalTo("abc"))
        assertThat(extension, equalTo(".jpg"))
    }

    @Test
    fun testFileNameTwoDot() {
        val (fileName, extension) = getValidFileNameAndExtension("a.b.c")
        assertThat(fileName, equalTo("a.b"))
        assertThat(extension, equalTo(".c"))
    }

    @Test
    fun `test create temp file - dot`() {
        val fileNameAndExtension = getValidFileNameAndExtension("a.b.c")
        File.createTempFile(fileNameAndExtension.fileName, fileNameAndExtension.extensionWithDot)
    }

    @Test
    fun `test create temp file - too short`() {
        // createTempFile fails with a 2-character filename
        val invalidTempFile = getValidFileNameAndExtension("ok.computer")
        assertThrows<IllegalArgumentException> {
            File.createTempFile(invalidTempFile.fileName, invalidTempFile.extensionWithDot)
        }
        val valid = invalidTempFile.renameForCreateTempFile()
        File.createTempFile(valid.fileName, valid.extensionWithDot)
    }

    @Test
    fun `string representation is unchanged`() {
        val underTest = getValidFileNameAndExtension("file.ext")
        assertThat("toString()", underTest.toString(), equalTo("file.ext"))
    }

    @Test
    fun `extension replacement dot handling`() {
        // I felt that requiring the '.' prefix was unintuitive and would lead to errors
        // so both the missing and valid cases are handled

        val underTest = getValidFileNameAndExtension("file.ext")

        assertThat("replace extension, no .", underTest.replaceExtension(extension = "file").toString(), equalTo("file.file"))
        assertThat("replace extension, with .", underTest.replaceExtension(extension = ".file").toString(), equalTo("file.file"))
    }

    /**
     * Allows destructuring of a [FileNameAndExtension]
     *
     * ```kotlin
     * val (fileName, extension) = getValidFileNameAndExtension("a.b.c")
     * ```
     *
     * Destructuring doesn't work on a `FileNameAndExtension?`
     * */
    private fun getValidFileNameAndExtension(input: String) = assertNotNull(FileNameAndExtension.fromString(input))

    @Test
    @Throws(Exception::class)
    fun `withFileNameSafe - valid child file returns File`() {
        val parentDir = temporaryDirectory.newFolder("parent")
        val result = parentDir.withFileNameSafe("child.txt")
        assertEquals(File(parentDir, "child.txt"), result)
    }

    @Test
    @Throws(Exception::class)
    fun `withFileNameSafe - nested valid child returns File`() {
        val parentDir = temporaryDirectory.newFolder("parent")
        val result = parentDir.withFileNameSafe("subdir/nested/file.txt")
        assertEquals(File(parentDir, "subdir/nested/file.txt"), result)
    }

    @Test
    @Throws(Exception::class)
    fun `withFileNameSafe path traversal with dotdot throws SecurityException`() {
        val parentDir = temporaryDirectory.newFolder("parent")
        try {
            parentDir.withFileNameSafe("../outside.txt")
            fail("Expected SecurityException for path traversal")
        } catch (e: SecurityException) {
            assertTrue(e.message!!.contains("traversal"))
        }
    }

    @Test
    @Throws(Exception::class)
    fun `withFileNameSafe double dotdot traversal throws SecurityException`() {
        val parentDir = temporaryDirectory.newFolder("parent")
        try {
            parentDir.withFileNameSafe("../foo/../../outside.txt")
            fail("Expected SecurityException for path traversal")
        } catch (e: SecurityException) {
            assertTrue(e.message!!.contains("traversal"))
        }
    }

    @Test
    @Throws(Exception::class)
    fun `withFileNameSafe child named same as parent in different location throws SecurityException`() {
        val parentDir = temporaryDirectory.newFolder("parent")
        val siblingDir = temporaryDirectory.newFolder("sibling_parent")
        // Create a file at sibling_dir/parent to test the boundary case
        File(siblingDir, "parent").mkdirs()
        try {
            parentDir.withFileNameSafe("../sibling_parent/parent/file.txt")
            fail("Expected SecurityException for same-name traversal")
        } catch (e: SecurityException) {
            assertTrue(e.message!!.contains("traversal"))
        }
    }

    @Test
    @Throws(Exception::class)
    fun `withFileNameSafe - file named like parent but not traversing stays in place`() {
        val parentDir = temporaryDirectory.newFolder("parent")
        // A file literally named "parent" inside the directory should be valid
        val result = parentDir.withFileNameSafe("parent")
        assertEquals(File(parentDir, "parent"), result)
    }

    @Test
    @Throws(Exception::class)
    fun `withFileNameSafe sibling prefix parentfoo is rejected`() {
        val root = temporaryDirectory.root
        val parentDir = File(root, "parent").apply { mkdirs() }
        File(root, "parentfoo").mkdirs()
        try {
            parentDir.withFileNameSafe("../parentfoo/evil.txt")
            fail("Expected SecurityException for sibling-prefix traversal")
        } catch (e: SecurityException) {
            assertTrue(e.message!!.contains("traversal"))
        }
    }

    @Test
    @Throws(Exception::class)
    fun `withFileNameSafe symlink escape throws SecurityException`() {
        assumeTrue(
            "Symlink test requires POSIX file attributes",
            FileSystems.getDefault().supportedFileAttributeViews().contains("posix"),
        )
        val parentDir = temporaryDirectory.newFolder("symlinkParent")
        val outsideDir = temporaryDirectory.newFolder("symlinkOutside")
        val target = File(outsideDir, "secret.txt").apply { writeText("secret") }
        val link = File(parentDir, "link").toPath()
        try {
            Files.createSymbolicLink(link, target.toPath())
        } catch (e: UnsupportedOperationException) {
            assumeTrue("Could not create symlink: ${e.message}", false)
        } catch (e: IOException) {
            assumeTrue("Could not create symlink: ${e.message}", false)
        }
        try {
            parentDir.withFileNameSafe("link")
            fail("Expected SecurityException for symlink escape")
        } catch (e: SecurityException) {
            assertTrue(e.message!!.contains("traversal"))
        }
    }
}
